doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,796 @@
+are not aware of each other.
+
+
+**Notes:**
+
+The proposed ``CastingImpl`` this designed to be compatible with the
+``UFuncImpl`` proposed in NEP YY.
+While initially it will be a distinct object, the aim is that ``CastingImpl``
+can be a subclass of ``UFuncImpl``.
+Once this happens, this will naturally allow the use of a ``CastingImpl`` to
+pass around a specialized casting function directly if so wished.
+
+In the future, we may considering adding a way to spell out that specific
+casts are known to be *not* possible.
+
+
+C-Side API
+^^^^^^^^^^
+
+.. note:: At the time of writing, this API is a general design goal.
+          Due to the size of the proposed changes, details and names will be
+          in flux and updated as necessary.
+
+DType creation
+""""""""""""""
+
+As already mentioned in NEP 41, the interface to define new DataTypes in C
+is modelled after the limited API in Python, the above mentioned slots,
+and some additional necessary information will thus be passed within a slots
+struct and identified by ``ssize_t`` integers::
+
+    static struct PyArrayMethodDef slots[] = {
+        {NPY_dt_method, method_implementation},
+        ...,
+        {0, NULL}
+    }
+
+    typedef struct{
+      PyTypeObject *typeobj;    /* type of python scalar */
+      int flexible;             /* Is the dtype flexible? */
+      int abstract;             /* Is the dtype abstract? */
+      int flags;                /* Currently only the "needs API" flag */
+      /* NULL terminated CastingImpl; is copied and references are stolen */
+      CastingImpl *castingimpls[];
+      PyType_Slot *slots;
+    } PyArrayDTypeMeta_Spec;
+
+    PyObject* PyArray_InitDTypeMetaFromSpec(
+            PyArray_DTypeMeta *user_dtype, PyArrayDTypeMeta_Spec *dtype_spec);
+
+all of this information will be copied during instantiation.
+
+The proposed method slots are (prepended with ``NPY_dt_``), these are
+detailed above and given here for summary:
+
+* ``is_native(self) -> {0, 1}``
+* ``ensure_native(self) -> dtype``
+* ``default_descr(self) -> dtype`` (return must be native and should normally be a singleton)
+* ``get_sort_function(self, NPY_SORTKIND sort_kind, void **out_sortfunction) -> success {-1, 0}``.
+  ``out_sortfunction`` may be ``NULL`` if the sort kind is unknown or not implemented.
+* ``setitem(self, char *item_ptr, PyObject *value) -> {-1, 0}``
+* ``getitem(self, char *item_ptr, PyObject (base_obj) -> object or NULL``
+* ``discover_descr_from_pyobject(cls, PyObject) -> dtype or NULL``
+* ``common_dtype(cls, other) -> DType, NotImplemented, or NULL``
+* ``common_instance(self, other) -> dtype or NULL``
+
+If not set, most slots are filled with slots which either error or defer automatically.
+Non-flexible dtypes do not have to implement:
+
+* ``discover_descr_from_pyobject`` (uses ``default_descr`` instead)
+* ``common_instance`` (uses ``default_descr`` instead)
+* ``ensure_native`` (uses ``default_descr`` instead)
```

Comment:
Ahh! Yes, we could do that. Hmmmmm.

Reply:
On the one hand i agree (from the Python API its a nice choice for sure), on the other hand, it does seem a bit clumsy from C maybe? Note that in the end I think I want to store the flag in either case, since, if it is valuable at all, lookup has to be *very* quick in C, and subclass checking is slower (very slow for abstract classes, although I could hand optimize that).